api/src/main/java/jakarta/data/Order.java

```diff
@@ -28,11 +28,11 @@
 /**
  * <p>Requests sorting on various entity attributes.</p>
  *
- * <p>{@code Order} can be optionally specified as a
- * parameter to a repository find method in any of the positions
- * that are after the query parameters, or it can be used
- * to obtain a {@link PageRequest} that is similarly
- * specified as a parameter to a repository find method.</p>
+ * <p>A query method of a repository may have a parameter of type
+ * {@code Order} if its return type indicates that it may return
+ * multiple entities. The parameter of type {@code Order} must occur
+ * after the method parameters representing regular parameters of
+ * the query itself.</p>
```

Comment:
Restricting this to multiple return type will prevent users from writing queries like this,
```
Optional<Employee> findFirstByJobTitle(String jobTitle, Order<Employee> sortBy);

highestPaid = employees.findFirstByJobTitle("Software Engineer", Order.by(
                                            _Employee.salary.desc(),
                                            _Employee.id.desc()));
```
and is contradictory to the description of NonUniqueResultException which instructs users to take this sort of approach with singular result to avoid the exception if they really want a single entity returned.

```suggestion
 * <p>A query method of a repository that performs a find operation
 * may have a parameter of type {@code Order}. The parameter of type
 * {@code Order} must occur after the method parameters representing
 * regular parameters of the query itself.</p>
```

Reply:
Yeah, I mean, OK, I guess I think that such a repository method has _extremely_ weak semantics.

Find the "first" X where "first" is something completely arbitrary determined by the caller.

I can't see myself ever writing a repository method like that TBH.

Reply:
In particular, `PageRequest` is pretty aesthetically-offensive for a query which returns a single entity.

Reply:
> In particular, `PageRequest` is pretty aesthetically-offensive for a query which returns a single entity.

I didn't mean to make the comment on `PageRequest`, only `Order` and `Sort`.

Reply:
If you _really_ want this, 63686513872418b11f76ab5e39c4664b51f5b307 rolls back that change for everything except `PageRequest`.

Reply:
> I didn't mean to make the comment on `PageRequest`, only `Order` and `Sort`.

What about `Limit`?

Reply:
>  I were going to write this method, I would rather write it as

Yeah, yeah, that's the point I was trying to make. You would want the ordering to be specified statically, I believe.

> Actually, I'd rather use `@Find` and name it "highestPaid" instead

Yeah, I agree, it would be _very_ nice to be able to write:

```java
@Find @First
@OrderBy(value = _Employee.SALARY, descending = true)
@OrderBy(value = _Employee.ID)
Optional<Employee> highestPaid(String jobTitle);
```

Where `@First` would suppress the exception that you would otherwise receive by adding `limit 1` to the query, and so you would just get the top result.

At some point we should add that.

Reply:
That's a good idea separating out `@First` from `@Find` because then it could also apply to `@Query`.  Just jotting this down for future reference.

For now, having thought about this some more, I don't like how `NonUniqueResultException` is pointing at a solution that only currently works with the Query by Method Name pattern (and even there it is awkward to use) and not the other patterns.  Let's go with what you proposed in this pull to only allows these for multiple return type and then we can update the recommendation in `NonUniqueResultException` to be to use a multiple return type.  I'll locate my review comments and mark them resolved with this.

Reply:
Oops, I was too late for that. I'll just approve this and if you want to put the language back limiting to multiple, go ahead.

Reply:
So, Nathan, what was your final conclusion here? 

Should I roll back https://github.com/jakartaee/data/pull/530/commits/63686513872418b11f76ab5e39c4664b51f5b307 or no?

Reply:
> So, Nathan, what was your final conclusion here?
> 
> Should I roll back [6368651](https://github.com/jakartaee/data/commit/63686513872418b11f76ab5e39c4664b51f5b307) or no?

Either way is fine with me.  This is a very minor detail and so I chose to approve and merge the pull yesterday (Otavio had already approved it) to ensure we could get the rest of your changes in before the March 8 cutoff.

Reply:
OK, well, hmmmm, I guess my preference would be to put the restriction back in, if only because it's really easy to remove a restriction like this later if we decide it was a mistake.

Reply:
> I guess my preference would be to put the restriction back in, if only because it's really easy to remove a restriction like this later if we decide it was a mistake.

That's a very good point. My only hesitation is that we passed the deadline, but this is a gray area because it could be framed as reverting a commit that shouldn't have gone in rather than adding something.

Reply:
I don't think rolling back a single commit, with the effect that there are fewer things to test and implement, is going to interfere with finishing the work on JDQL.